src/pyodide/internal/_workers.py

```diff
@@ -942,6 +990,36 @@ def wrapped_init(self, *args, **kwargs):
 
     cls.__init__ = wrapped_init
 
+def _wrap_workflow_step(cls):
+    run_fn = getattr(cls, "on_run", None)
+    if run_fn is None:
+        return
+
+    # Only patch `on_run` for subclasses of WorkflowEntrypoint.
+    is_workflow_subclass = any(
+        base.__name__ == "WorkflowEntrypoint" for base in cls.__mro__[1:]
+    )
+
+    if not is_workflow_subclass:
+        # Not a workflow subclass, so don't wrap `on_run`.
+        return
+
+    if asyncio.iscoroutinefunction(run_fn):
+        async def wrapped_run(self, *args, **kwargs):
+            if len(args) > 1:
+                args = (python_from_rpc(args[0]), _WorkflowStepWrapper(args[1])) + args[2:]
+
+            return await run_fn(self, *args, **kwargs)
```

Comment:
To reduce duplication, maybe we can always define an async wrapper and do the following:
```suggestion
            result = run_fn(self, *args, **kwargs)
            if iscoroutine(result):
                result = await result
            return result
```